cluster-autoscaler/cloudprovider/mcm/mcm_manager.go

```diff
@@ -623,47 +600,25 @@ func (m *McmManager) GetMachineDeploymentNodeTemplate(machinedeployment *Machine
 	)
 
 	switch machineClass.Kind {
-	case kindAWSMachineClass:
-		mc, err := m.awsMachineClassLister.AWSMachineClasses(m.namespace).Get(machineClass.Name)
-		if err != nil {
-			return nil, fmt.Errorf("Unable to fetch AWSMachineClass object %s, Error: %v", machineClass.Name, err)
-		}
-		awsInstance, exists := aws.InstanceTypes[mc.Spec.MachineType]
-		if !exists {
-			return nil, fmt.Errorf("Unable to fetch details for VM type %s", mc.Spec.MachineType)
-		}
-		instance = instanceType{
-			InstanceType: awsInstance.InstanceType,
-			VCPU:         awsInstance.VCPU,
-			MemoryMb:     awsInstance.MemoryMb,
-			GPU:          awsInstance.GPU,
-		}
-		region = mc.Spec.Region
-		zone = getZoneValueFromMCLabels(mc.Labels)
-	case kindAzureMachineClass:
-		mc, err := m.azureMachineClassLister.AzureMachineClasses(m.namespace).Get(machineClass.Name)
-		if err != nil {
-			return nil, fmt.Errorf("Unable to fetch AzureMachineClass object %s, Error: %v", machineClass.Name, err)
-		}
-		azureInstance, exists := azure.InstanceTypes[mc.Spec.Properties.HardwareProfile.VMSize]
-		if !exists {
-			return nil, fmt.Errorf("Unable to fetch details for VM type %s", mc.Spec.Properties.HardwareProfile.VMSize)
-		}
-		instance = instanceType{
-			InstanceType: azureInstance.InstanceType,
-			VCPU:         azureInstance.VCPU,
-			MemoryMb:     azureInstance.MemoryMb,
-			GPU:          azureInstance.GPU,
-		}
-		region = mc.Spec.Location
-		if mc.Spec.Properties.Zone != nil {
-			zone = mc.Spec.Location + "-" + strconv.Itoa(*mc.Spec.Properties.Zone)
-		}
 	case kindMachineClass:
 		mc, err := m.machineClassLister.MachineClasses(m.namespace).Get(machineClass.Name)
 		if err != nil {
 			return nil, fmt.Errorf("Unable to fetch %s for %s, Error: %v", kindMachineClass, machineClass.Name, err)
 		}
+
+		if nodeTemplateAttributes := mc.NodeTemplate; nodeTemplateAttributes != nil {
+			klog.V(1).Infof("Generating node template using nodeTemplate from MachineClass %s", machineClass.Name)
+			instance = instanceType{
+				InstanceType: nodeTemplateAttributes.InstanceType,
+				VCPU:         nodeTemplateAttributes.Capacity["cpu"],
+				Memory:       nodeTemplateAttributes.Capacity["memory"],
+				GPU:          nodeTemplateAttributes.Capacity["gpu"],
+			}
+			region = nodeTemplateAttributes.Region
+			zone = nodeTemplateAttributes.Zone
```

Comment:
In case lets say cpu is not specified , then it will assigned as `0` over here, and that would actually lead to a wrong nodeTemplate for the machinedeployment ... same for zone, region not provided cases...

Though it won't be a problem as CA won't see a nodegroup with zero resources as a node fit to scale up and will scale up another node grp....

But I think just logging that resources are not present in MC from CA side would be ok for now, because otherwise we won't get to know of missing resources in MC until the machinecontroller tells in validation.
WDYT?

Reply:
Hmm, makes sense. I will introduce the change.